### PR TITLE
Update visualization auto-selection logic

### LINKED ISSUE #11675: Visualization type cannot be saved when is only one result
**Describe the bug**
Since updating from 0.33 to 0.34, on sql questions with only 1 row as result I cannot save the visualization type. Every time I refresh the question it reverts back to table visualization and makes a new question.
If I have more than 1 row as a result the visualization works. If I downgrade to 0.33.x it works again.

**To Reproduce**
1.Create a question with 1 row as result
2.Set visualization to bar
4. Save
3. Reload question

**Expected behavior**
The visualization should be saved.

**Screenshots**
Same issue even when I save before reload. When I open the question it automatically creates a new question from the old one with a table visualization. 
Visualization error

**Information about your Metabase Installation:**

```
{
  "browser-info": {
    "language": "en-VI",
    "platform": "Win32",
    "userAgent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/79.0.3945.117 Safari/537.36",
    "vendor": "Google Inc."
  },
  "system-info": {
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "11.0.5+10-post-Ubuntu-0ubuntu1.118.04",
    "java.vendor": "Private Build",
    "java.vendor.url": "Unknown",
    "java.version": "11.0.5",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "11.0.5+10-post-Ubuntu-0ubuntu1.118.04",
    "os.name": "Linux",
    "os.version": "4.15.0-74-generic",
    "user.language": "en",
    "user.timezone": "Europe/Bucharest"
  },
  "metabase-info": {
    "databases": [
      "postgres"
    ],
    "hosting-env": "unknown",
    "application-database": "mysql",
    "run-mode": "prod",
    "version": {
      "date": "2019-12-19",
      "tag": "v0.34.0",
      "branch": "release-0.34.x",
      "hash": "1335215"
    },
    "settings": {
      "report-timezone": null
    }
  }
}
```

### LINKED ISSUE #11960: Native question will reset visualization to Table unless there's data from the beginning
When making any change to a question, even one that doesn’t require Saving - like changing the Question description - the Display Option changes to Table.
It seems that Metabase always tries to infer the questions data result, but after any change it is like there is no sample data, so it loses any Display Option applied before.

**Note:** when we manually switch back to the desired Display Option, the previously used formatting (like column names, number formatting, column order, axis) are restored - they are not lost (gladly).

Metabase version - v0.34.1 - 2020-01-13

### LINKED ISSUE #12361: Visualization type gets reset to "Table" (different scenarios)
**Describe the bug**
Getting an error in a SQL query (eg: a column name changed) makes it be back to "table" visualization mode.

**Logs**
Nothing special

**To Reproduce**
1. Go to a SQL (edit:) or query-builder question that is a graph
2. Change the query to make it bug, for instance having a wrong column name
3. Play the query to get the results, see the error
4. Fix the error
5. Play the query, you're in table vizualisation mode.

[edit:] Other variant:
1. Create quickly a graph from a table
2. Refresh the tab, duplicate it, have your tab suspended by a browser plugin, or **share the link to a coworker**
3. The visualization is back to table mode

You need to get back in the proper visualization mode to get everything back again.

**Expected behavior**
Don't get stuck in the table visualization mode just because there is an error. Stay on the current mode, but switch to "data" view. instead of visualization view.

**Screenshots**

**Information about your Metabase Installation:**
```
{
  "browser-info": {
    "language": "en-US",
    "platform": "MacIntel",
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/81.0.4044.92 Safari/537.36",
    "vendor": "Google Inc."
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "1.8.0_242-heroku-b08",
    "java.vendor": "Oracle Corporation",
    "java.vendor.url": "http://java.oracle.com/",
    "java.version": "1.8.0_242-heroku",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "25.242-b08",
    "os.name": "Linux",
    "os.version": "4.4.0-1062-aws",
    "user.language": "en",
    "user.timezone": "Etc/UTC"
  },
  "metabase-info": {
    "databases": [
      "postgres",
      "redshift",
      "h2"
    ],
    "hosting-env": "heroku",
    "application-database": "postgres",
    "application-database-details": {
      "database": {
        "name": "PostgreSQL",
        "version": "9.6.16"
      },
      "jdbc-driver": {
        "name": "PostgreSQL JDBC Driver",
        "version": "42.2.8"
      }
    },
    "run-mode": "prod",
    "version": {
      "tag": "v0.35.2",
      "date": "2020-04-10",
      "branch": "release-0.35.x",
      "hash": "f3b2857"
    },
    "settings": {
      "report-timezone": "UTC"
    }
  }
}
```

**Severity**
Severity is high. It's not been here forever, I think it's one of the secret gifts of the 0.34 :-) (but not sure, might be 0.35). Impact:
- when iterating on a complicated query with graphs, a few unneeded clicks and a poor UX
- [**more important**]: when fixing an existing query made wrong by underlying data changes, it's impossible to fix the question entirely because you have to "guess" if there was a visualization before and which one.
- [edit:] [**most important**]: you cannot share a question with a link to a coworker.

### PR DESCRIPTION
**Pull request by paulrosenzweig**
This PR tweaks the auto-selection behavior in a few ways:
1. The "locked" status is persisted in the URL, so you can send an unsaved question link without the recipient reverting to the default viz type. (Fixes #12361)
2. Don't run auto-selection unless the card is dirty. (Fixes #11675)
3. Don't unlock display when a (presumably temporary) dataset is returned with zero rows. (Fixes #11960)

## How it works now
(This is mostly unchanged from before, but just for posterity.)

When you create a new question and start editing it, we auto select the right display type based on the query and the data. If you explicitly choose a type in the left sidebar, we lock in that choice as you update the query. We only override it if the visualization is no longer "sensible" for the new data (e.g. a Gauge chart when there are multiple rows).

Saved questions are locked to a display type as though you selected it. Additionally, we _won't_ override even if the type isn't sensible until you edit the question. Saving a question shows that you want to preserve how it currently appears.

Here's the same info in bullets: 
* editing a new question -> we take the display from auto-selection
* editing a question after picking vis type -> use the display if sensible
* viewing an unchanged saved question ->  always use the display
* editing a saved question -> use the display if sensible

---

Separately from those fixes. I refactored how we lock display on a card. I previously left a comment that the data was "essentially a boolean flag". I changed it so it actually is a boolean. (No idea why I didn't before)

### CODE REVIEW

**Comment by flamber** [comment]
@paulrosenzweig Much better behavior than before :tada: :+1: Also fixes #12014 and https://github.com/metabase/metabase/issues/11218#issuecomment-583115514
1. Fixed.
2. It's still resetting to table for some viz, when there's only one row, when the question is unsaved. That seems like a minor annoyance compared to before.
Tested with Bar viz with following query and modifying the limit between 0 and 2:
```
SELECT YEAR(CREATED_AT), SUM(PRICE) FROM PRODUCTS GROUP BY YEAR(CREATED_AT) LIMIT 2
```
Doing something similar with Map viz doesn't reset to Table:
```
SELECT STATE, COUNT(*) FROM PEOPLE GROUP BY STATE LIMIT 1
```
3. Fixed.

File: frontend/src/metabase/query_builder/actions.js [hunk 1]
```diff
@@ -881,6 +885,8 @@ export const runQuestionQuery = ({
       ? question.isDirtyComparedTo(originalQuestion)
       : true;
 
+    // const cardIsDirty = getIsQuestionDirtyFromOriginal(getState(), question)
```
Comment (kdoh) [inline frontend/src/metabase/query_builder/actions.js:888]: Is this not needed anymore?

Comment (paulrosenzweig) [reply frontend/src/metabase/query_builder/actions.js:888]: nope. Thanks!

**Comment by paulrosenzweig** [comment]
@flamber thanks for pointing that out. I just played around with it a bit more. That's expected for the current rules. Whether those rules are right is a separate question. 😄

For context, it's switching to a table because the bar chart isn't "sensible" for a single row of data. (Maps however are considered sensible with a single row.) It'd be the same as switching to a table after selecting a gauge if a second row appeared.

I think I agree with those chart-specific sensibility criteria. Bar charts should be used to compare at least two quantities. Maps work just as well to display single pin or region as they do for two. 

Our vis selection is nice when it enables you to pull down new data and see a better chart for the data's shape. However, it gets annoying when you're exploring the data and have your deliberate choice of chart overruled.

It's pretty common to be exploring data and accidentally get a result with zero rows. This PR adds a special case for zero-row result sets to ignore the normal sensibility logic. This prevents unlocking the display selection due to a presumably temporary query change.

Should we do the same thing for a single row? I think it's less likely that you accidentally get one row compared to accidentally getting zero rows, but I also don't see much of a downside. Switching a single row to a table isn't a substantially better choice here. I think I'm in favor and just pushed the change to try it out.

**Comment by flamber** [comment]
@paulrosenzweig :+1: On the last part, I think that would fix issues like 8623 (not linking)

**Comment by mazameli** [comment]
Messed around with this a bit and didn't notice anything weird. @flamber, were there remaining problems you saw? Do we need a code review from @tlrobinson, Paul?

**Comment by paulrosenzweig** [comment]
@mazameli the code didn't change much. I'm good merging if the behavior looks good to you guys.

**Review by flamber — APPROVED** [review]
Paul made it so even single row doesn't revert back to table, so I'm good with this from a non-code perspective. Would be nice to have this merged now, so we have some time to catch any weirdness.
Merge to `master` instead perhaps?

**Review by kdoh — APPROVED** [review]

**Pull request merged**